ray-curator/ray_curator/stages/deduplication/fuzzy/connected_components.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from typing import TYPE_CHECKING, Any
+
+import cudf
+import cugraph
+from cugraph.dask.comms.comms_wrapper import init_subcomms as c_init_subcomms
+from loguru import logger
+from pylibcugraph import GraphProperties, MGGraph, ResourceHandle
+from pylibcugraph import weakly_connected_components as pylibcugraph_wcc
+
+from ray_curator.backends.experimental.utils import RayStageSpecKeys
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.deduplication.id_generator import (
+    CURATOR_DEDUP_ID_STR,
+)
+from ray_curator.stages.deduplication.io_utils import DeduplicationIO
+from ray_curator.stages.resources import Resources
+from ray_curator.tasks.file_group import FileGroupTask
+from ray_curator.utils.file_utils import delete_dir, get_fs, is_not_empty
+
+if TYPE_CHECKING:
+    from ray_curator.backends.base import WorkerMetadata
+
+
+class ConnectedComponentsStage(ProcessingStage[FileGroupTask, FileGroupTask], DeduplicationIO):
+    def __init__(
+        self,
+        output_dir: str,
+        source_column: str = f"{CURATOR_DEDUP_ID_STR}_x",
+        destination_column: str = f"{CURATOR_DEDUP_ID_STR}_y",
+        read_kwargs: dict | None = None,
+        write_kwargs: dict | None = None,
+    ):
+        """
+        Args:
+            output_dir: The directory to write the resulting connected components to.
+            source_column: The column name containing the document ids of the source of the edge.
+            destination_column: The column name containing the document ids of the destination of the edge.
+            read_kwargs: Keyword arguments to pass for reading the input files.
+            write_kwargs: Keyword arguments to pass for writing the output files.
+        """
+        # Initialize parent classes
+        ProcessingStage.__init__(self)
+        DeduplicationIO.__init__(self, id_generator=None)
+
+        self.source_column = source_column
+        self.destination_column = destination_column
+        self.read_kwargs = read_kwargs if read_kwargs is not None else {}
+        self.write_kwargs = write_kwargs if write_kwargs is not None else {}
+
+        self._name = self.__class__.__name__
```

Comment:
Why this vs just "ConnectedComponents"?

Reply:
I've generally been following the convention of naming this `MinHashStage, LSHStage etc.` instead of just ConnectedComponents or LSH. I haven't looked at the other stage names in too much detail but don't have strong opinions. Happy to have it be ConnectedComponents